### PR TITLE
feat: add new prompt best practices section

### PR DESCRIPTION
**Pull request by PeriniM**
```mermaid
sequenceDiagram
    participant User
    participant LangSmith
    participant Intermediary Service
    participant GitHub Repo

    User->>LangSmith: Commits new prompt version
    activate LangSmith
    LangSmith->>Intermediary Service: Fires Webhook (sends prompt data)
    deactivate LangSmith
    activate Intermediary Service
    Intermediary Service->>GitHub Repo: Commits new prompt file via API
    activate GitHub Repo
    GitHub Repo-->>Intermediary Service: Confirms commit
    deactivate GitHub Repo
    Intermediary Service-->>LangSmith: Responds with success (optional)
    deactivate Intermediary Service
``` 

```mermaid
gitGraph
    commit id: "v1.0"
    branch develop
    checkout develop
    commit id: "feat: new idea"
    commit id: "fix: refine idea" tag: "dev"
    checkout main
    merge develop id: "Release v2.0" tag: "staging" tag: "prod"
    checkout develop
    commit id: "feat: another feature"
``` 

```mermaid
graph TD
    subgraph Trigger
        A[Commit New Prompt Version];
    end

    subgraph "CI: Automated Testing"
        B(Step 1: Linting & Validation);
        C(Step 2: Unit Tests);
        D(Step 3: Run Evals on Dataset);
    end

    subgraph "CD: Automated Deployment"
        E{Evals Pass & Score Improves?};
        F[Auto-Promote to 'staging'];
        G[Fail & Alert Team];
        H[Manual Review of Staging];
        I[Promote to 'prod'];
    end

    A --> B;
    B --> C;
    C --> D;
    D --> E;
    E -- Yes --> F;
    E -- No --> G;
    F --> H;
    H --> I;
```

### CODE REVIEW

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+### Evaluating prompt changes before promoting to production
+
+A typical pipeline consists of several automated stages that run in sequence. If any stage fails, the pipeline stops and notifies the team.
+
+![CICD Diagram](./static/prompt-best-practice-cicd.png)
+
+- **Stage 1: Trigger**
+  The pipeline starts automatically when a new prompt version is created.
+  - **How:** This can be a `git push` to your main branch or, more powerfully, a **webhook** triggered from LangSmith on every new prompt commit.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:71]: I'm not sure what "more powerfully" means here? Is there a word that is more descriptive about why webhooks are better?

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+- **Stage 2: Linting & unit tests**
+  This stage performs quick, low-cost checks.
+  - **Linting:** A simple script checks for basic syntax. For example, does the prompt template contain all the required input variables (e.g., `{question}`, `{context}`)?
+  - **Unit Tests:** These verify the _structure_ of the output, not the quality. You can use a framework like `pytest` to make a few calls to the new prompt and assert things like:
+    - "Does the output always return valid JSON?"
+    - "Does it contain the expected keys?"
+    - "Is the list length correct?"
+- **Stage 3: Quality evaluation**
+  This is the heart of the pipeline. The new prompt version is run against your evaluation dataset to ensure it meets quality standards.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:80]:
```suggestion
  The new prompt version is run against your evaluation dataset to ensure it meets quality standards.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  - **How:** This can be done in a few ways:
+    - **Programmatic SDK Evals:** For checks against known ground truths, a script can use the LangSmith SDK's `evaluate` [function](https://docs.smith.langchain.com/evaluation#7-run-and-view-results). This executes the new prompt against every example in your dataset, and the results are automatically scored by your chosen programmatic evaluators (e.g., for JSON validity, string matching).
+    - **Advanced Qualitative Evals with `openevals`:** For more nuanced quality checks (like helpfulness, style, or adherence to complex instructions), you can leverage the `openevals` library. This library integrates directly with `pytest` and allows you to define sophisticated, "LLM-as-a-judge" evaluations. You can create tests that use another LLM to score the quality of your prompt's output. Thanks to the [LangSmith integration](https://github.com/langchain-ai/openevals?tab=readme-ov-file#langsmith-integration), all these evaluation runs are automatically traced and can be visualized in LangSmith, giving you a detailed view of the results.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:83]:
```suggestion
    - **Advanced Qualitative Evals with `openevals`:** For more nuanced quality checks (like helpfulness, style, or adherence to complex instructions), you can leverage the `openevals` library. This library integrates directly with `pytest` and allows you to define sophisticated, "LLM-as-a-judge" evaluations. You can create tests that use another LLM to score the quality of your prompt's output. The [LangSmith integration](https://github.com/langchain-ai/openevals?tab=readme-ov-file#langsmith-integration) automatically traces and visualizes all these evaluation runs in LangSmith, which provides a detailed view of the results.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  - **The Check:** The pipeline then compares the new prompt's aggregate evaluation scores (e.g., average correctness, helpfulness score, latency, cost) against the current production prompt's scores.
+- **Stage 4: Continuous deployment (promotion)**
+  Based on the evaluation results, the prompt is automatically promoted.
+  - **Pass/Fail Logic:** The pipeline checks if the new prompt is "better" than the current one (e.g., higher correctness score, no drop in helpfulness, within cost budget).
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:87]:
```suggestion
  - **Pass/Fail Logic:** The pipeline checks if the new prompt is "better" than the current one based on criteria such as higher correctness score, no drop in helpfulness, within cost budget.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  - **Promotion to `staging`:** If it passes, a script uses the LangSmith SDK to move the `staging` tag to this new commit hash. Your staging application, which pulls the `your-prompt:staging` tag, will automatically start using the new prompt.
+  - **Promotion to `prod`:** This is often a **manual step**. After the team has reviewed the performance in the staging environment and is confident in the results, a team member can manually move the `prod` tag in the LangSmith UI. This final human check is a crucial safety measure.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:89]: "This is often a manual step" Is there a way for this to be automatic? We then say that it is crucial to include a human check — so wondering if it should ever be automatic?

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+## Syncing prompts in production
+
+For better version control, collaboration, and integration with your existing CI/CD pipelines, synchronizing your LangSmith prompts with an external source code repository will give you a full commit history alongside your application code.
+
+### Best practice: use webhooks for synchronization
+
+The most effective way to automate this is by using webhooks. You can configure LangSmith to send a notification to a service every time a new version of a prompt is saved. This creates a seamless bridge between the user-friendly prompt editing environment in LangSmith and your rigorous version control system.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:97]:
```suggestion
The most effective way to automate this is by using webhooks. You can configure LangSmith to send a notification to a service every time a new version of a prompt is saved. This creates a seamless bridge between the user-friendly prompt editing environment in LangSmith and your version control system.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+### How to implement this with LangSmith
+
+LangSmith allows you to configure a webhook for your workspace that will fire on every prompt commit. You can point this webhook to your own service (like an AWS Lambda function or a small server) to handle the synchronization logic.
+
+**➡️ Learn more in the official documentation:** [Trigger a webhook on prompt commit](https://docs.smith.langchain.com/prompt_engineering/how_to_guides/trigger_webhook), [How to Sync Prompts with GitHub](https://docs.smith.langchain.com/prompt_engineering/tutorials/prompt_commit)
+
+## How to use a prompt in production without repeated API calls
+
+In a production application, making a network request to fetch a prompt every time a function is executed adds unnecessary latency and creates a dependency that could fail. The application should be resilient and fast.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:113]:
```suggestion
In a production application, making a network request to fetch a prompt every time a function is executed adds unnecessary latency and creates a dependency that could fail.
```
What do you think about dropping this sentence, I think you cover this well in the next paragraph.

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  - **Pros**: Extremely fast access with sub-millisecond latency; no additional infrastructure required.
+  - **Cons**: The cache is lost if the application restarts. Each instance of your application (if you have more than one server) will have its own separate cache, which could lead to brief inconsistencies when a prompt is updated.
+  - **Best for**: Single-instance applications, development environments, or applications where ultimate consistency across all nodes is not critical.
+- **2. Distributed caching**
+  This approach uses an external, centralized caching service that is shared by all instances of your application.
+  - **How it works**: Your application instances connect to a shared caching service like **Redis** or **Memcached**. When a prompt is needed, the application first checks the distributed cache. If it's not there (a "cache miss"), it fetches the prompt from LangSmith, stores it in the cache, and then uses it.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:131]:
```suggestion
  - **How it works**: Your application instances connect to a shared caching service like [Redis](https://redis.io/) or [Memcached](https://memcached.org/). When a prompt is needed, the application first checks the distributed cache. If it's not there (a "cache miss"), it fetches the prompt from LangSmith, stores it in the cache, and then uses it.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 9]
```diff
@@ -0,0 +1,134 @@
+---
+sidebar_label: Best Practices
+sidebar_position: 5
+---
+
+Here are our recommended best practices for managing your prompts. This guide will help you establish a robust workflow for developing, testing, and deploying prompts using LangSmith.
+
+## The prompt development lifecycle: iterating in the playground
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:8]:
```suggestion
## The prompt development lifecycle: iterate in the playground
```
Shall we provide a link to the Playground?

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+Before a prompt is versioned or deployed, it must be created. The best practice for prompt development is to treat it as an iterative, experimental process. The LangSmith Playground is the ideal environment for this initial "development" phase.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:10]:
```suggestion
We recommend treating  prompt development as an iterative, experimental process. The LangSmith Playground is the ideal environment for this initial "development" phase.
```
What do you think of this?

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+Using the Playground, you and your team can:
+
+- **Rapidly iterate on prompts:** Modify prompt templates on the fly and immediately see how the changes affect the output.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:14]:
```suggestion
- **Rapidly iterate on prompts:** Modify prompt templates and see how the changes affect the output immediately.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+- **Compare different LLMs:** Test the same prompt against various models (e.g., GPT-4o vs. Claude 3 Opus vs. Llama 3) side-by-side to find the best one for the job. This is crucial, as a prompt's effectiveness can vary significantly between models.
+- **Test with diverse inputs:** Easily run the prompt and model configuration against a range of different inputs to check for edge cases and ensure reliability.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:16]:
```suggestion
- **Test with diverse inputs:** Run the prompt and model configuration against a range of different inputs to check for edge cases and ensure reliability.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+- **Optimize the prompt**: Use the Prompt Canvas feature to use an LLM to improve your prompt
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:17]:
```suggestion
- **Optimize the prompt**: Use the [Prompt Canvas](./prompt_engineering/how_to_guides/prompt_canvas.mdx) feature to have an LLM improve your prompt.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+  ➡️ **See the blog post:** [LangChain Changelog](https://changelog.langchain.com/announcements/prompt-canvas-for-streamlined-prompt-engineering)
+- **Develop and test tool calling:** Configure tools and functions that the LLM can call, and test the full interaction within the Playground.
+- **Refine your app:** Run experiments directly against your dataset in the Playground, seeing changes in real time as you iterate on prompts. Share these experiments with teammates to get feedback and collaboratively optimize performance.
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:20]:
```suggestion
- **Refine your app:** Run experiments directly against your dataset in the Playground to see changes in real time as you iterate on prompts. Share experiments with teammates to get feedback and collaboratively optimize performance.
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+Once you are satisfied with a prompt and it’s configuration in the Playground, you can save it as a new commit to your prompt's history. While the Playground UI is great for experimentation, you can also create and update prompts programmatically for more automated workflows using the LangSmith SDK.
+
+**➡️ See the docs:** https://docs.smith.langchain.com/prompt_engineering/how_to_guides/manage_prompts_programatically
+
+**➡️ See the SDK reference:** [client.create_prompt](https://docs.smith.langchain.com/reference/python/client/langsmith.client.Client#langsmith.client.Client.create_prompt)
+
+## Managing prompts through different environments
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:28]:
```suggestion
## Manage prompts through different environments
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+Prompts are not static text; they are a fundamental component of your LLM application's logic, just like source code. A minor change can significantly alter an LLM's response or tool selection, making structured lifecycle management essential. The **LangSmith Prompt Hub** provides a central workspace to manage this complexity. This guide details the complete workflow for using the Hub to test prompts in development, validate them in staging, and deploy them confidently to production.
+
+## Dynamically updating application prompts based on prompt tags
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:32]:
```suggestion
## Update application prompts based on prompt tags
```
If we want to include "dynamically", I think we could thread that into the prose beneath this heading.

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+
+# In production, this code always fetches the stable prompt version currently tagged as "prod"
+prompt_prod = client.pull_prompt("your-prompt-name:prod")
+```
+
+**➡️ Learn more in the official documentation:** [Prompt Tags](https://docs.smith.langchain.com/prompt_engineering/how_to_guides/prompt_tags)
+
+![Git Diagram](./static/prompt-best-practice-git.png)
+
+### Evaluating prompt changes before promoting to production
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:63]:
```suggestion
### Evaluate prompt changes before promotion to production
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+  - **The Check:** The pipeline then compares the new prompt's aggregate evaluation scores (e.g., average correctness, helpfulness score, latency, cost) against the current production prompt's scores.
+- **Stage 4: Continuous deployment (promotion)**
+  Based on the evaluation results, the prompt is automatically promoted.
+  - **Pass/Fail Logic:** The pipeline checks if the new prompt is "better" than the current one (e.g., higher correctness score, no drop in helpfulness, within cost budget).
+  - **Promotion to `staging`:** If it passes, a script uses the LangSmith SDK to move the `staging` tag to this new commit hash. Your staging application, which pulls the `your-prompt:staging` tag, will automatically start using the new prompt.
+  - **Promotion to `prod`:** This is often a **manual step**. After the team has reviewed the performance in the staging environment and is confident in the results, a team member can manually move the `prod` tag in the LangSmith UI. This final human check is a crucial safety measure.
+
+## Syncing prompts in production
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:91]:
```suggestion
## Sync prompts in production
```

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+
+For better version control, collaboration, and integration with your existing CI/CD pipelines, synchronizing your LangSmith prompts with an external source code repository will give you a full commit history alongside your application code.
+
+### Best practice: use webhooks for synchronization
```
Comment (katmayb) [inline docs/prompt_engineering/tutorials/best_practices.mdx:95]: Here, and for the caching sub heading, you've included the "Best practice:" prefix to the subheading. Do you want to do the same for the prompt sections? For example, would you consider the "Evaluate prompt changes before promotion to production" the best practice for that section?

**Review by katmayb — COMMENTED** [review]
Hi @PeriniM, Tanushree sent me this PR to take a look through while I ramp up on LangSmith docs. I made some suggestions — I hope these are helpful. Lmk if you have any questions! cc @tanushree-sharma

**Comment by katmayb** [comment]
Since this is a best practices guide, maybe adding a link to this new page in the Get Started page ([here](https://docs.smith.langchain.com/#prompt-engineering)) would be helpful for users to review when looking at the prompt engineering links.

File: docs/prompt_engineering/tutorials/best_practices.mdx (hunk 19 above)
Comment (PeriniM) [reply docs/prompt_engineering/tutorials/best_practices.mdx:95]: Yes, let's add that!

File: docs/prompt_engineering/tutorials/best_practices.mdx [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+
+![CICD Diagram](./static/prompt-best-practice-sequence.png)
+
+### How to implement this with LangSmith
+
+LangSmith allows you to configure a webhook for your workspace that will fire on every prompt commit. You can point this webhook to your own service (like an AWS Lambda function or a small server) to handle the synchronization logic.
+
+**➡️ Learn more in the official documentation:** [Trigger a webhook on prompt commit](https://docs.smith.langchain.com/prompt_engineering/how_to_guides/trigger_webhook), [How to Sync Prompts with GitHub](https://docs.smith.langchain.com/prompt_engineering/tutorials/prompt_commit)
+
+## How to use a prompt in production without repeated API calls
```
[the remark this answers is not in the document]
Comment (PeriniM) [reply docs/prompt_engineering/tutorials/best_practices.mdx:111]: Yep, I agree otherwise we should have put it under the how_to section in the docs

File: docs/prompt_engineering/tutorials/best_practices.mdx (hunk 9 above)
Comment (PeriniM) [reply docs/prompt_engineering/tutorials/best_practices.mdx:8]: added link to playground before bullet points

File: docs/prompt_engineering/tutorials/best_practices.mdx (hunk 13 above)
Comment (PeriniM) [reply docs/prompt_engineering/tutorials/best_practices.mdx:17]: thanks for adding the link! Usually, .mdx is not necessary nor the relative path ./

File: docs/prompt_engineering/tutorials/best_practices.mdx (hunk 1 above)
Comment (PeriniM) [reply docs/prompt_engineering/tutorials/best_practices.mdx:71]: removed that expression, those methods are both valid.

File: docs/prompt_engineering/tutorials/best_practices.mdx (hunk 5 above)
Comment (PeriniM) [reply docs/prompt_engineering/tutorials/best_practices.mdx:89]: Can be automatic but we need to be sure our evaluation pipeline is trusthworthy and tackles most of the use cases. What about

This is often a manual step—a team member can move the prod tag in the LangSmith UI after reviewing performance in staging. However, if your evaluation pipeline is trustworthy and consistently reflects real-world performance, this step can be automated (e.g., canary deployments with rollback monitoring), and the prod tag can be advanced programmatically using the LangSmith SDK.
